core/string/ustring.h

```diff
@@ -83,92 +83,58 @@ class CharProxy {
 	}
 };
 
-/*************************************************************************/
-/*  Char16String                                                         */
-/*************************************************************************/
+template <typename T>
+class CharStringT {
+	friend class String;
 
-class Char16String {
-	CowData<char16_t> _cowdata;
-	static const char16_t _null;
+	CowData<T> _cowdata;
+	static constexpr T _null = 0;
 
 public:
-	_FORCE_INLINE_ char16_t *ptrw() { return _cowdata.ptrw(); }
-	_FORCE_INLINE_ const char16_t *ptr() const { return _cowdata.ptr(); }
+	_FORCE_INLINE_ T *ptrw() { return _cowdata.ptrw(); }
+	_FORCE_INLINE_ const T *ptr() const { return _cowdata.ptr(); }
 	_FORCE_INLINE_ int size() const { return _cowdata.size(); }
-	Error resize(int p_size) { return _cowdata.resize(p_size); }
+	_FORCE_INLINE_ Error resize(int p_size) { return _cowdata.resize(p_size); }
 
-	_FORCE_INLINE_ char16_t get(int p_index) const { return _cowdata.get(p_index); }
-	_FORCE_INLINE_ void set(int p_index, const char16_t &p_elem) { _cowdata.set(p_index, p_elem); }
-	_FORCE_INLINE_ const char16_t &operator[](int p_index) const {
+	_FORCE_INLINE_ T get(int p_index) const { return _cowdata.get(p_index); }
+	_FORCE_INLINE_ void set(int p_index, const T &p_elem) { _cowdata.set(p_index, p_elem); }
```

Comment:
```suggestion
	_FORCE_INLINE_ void set(int p_index, T p_elem) { _cowdata.set(p_index, p_elem); }
```
We can assume `T` will always be a simple type I'd say